### ISSUE TITLE
React: 3.0

### ISSUE THREAD

**Issue by cschroeter** [chakra-ui/ark#2330]
## Progress

Add `Context` parts and short `TypeDefinitions`:

- [x] Accordion
- [x] Avatar
- [x] Carousel
- [x] Checkbox
- [x] Clipboard
- [x] Collapsible
- [x] Color Picker
- [x] Combobox
- [x] Date Picker
- [x] Dialog
- [x] Editable
- [x] File Upload
- [x] Format
- [x] Hover Card
- [x] Menu
- [x] Number Input
- [x] Pagination
- [x] Pin Input
- [x] Popover
- [x] Progress
- [x] Radio Group
- [x] Rating Group
- [x] Segment Group
- [x] Select
- [x] Slider
- [x] Splitter
- [x] Switch
- [x] Tabs
- [x] Tags Input
- [x] Toast
- [x] Toggle Group
- [x] Tooltip
- [x] Tree View

### Added

- Exposed component-related types to keep imports clean and orderly.

```tsx
import { Avatar } from '@ark-ui/react';

export const Example = () => {
  // New: Use `Avatar` import to declare types.
  const handleLoadingStatusChange = (details: Avatar.StatusChangeDetails) => {
    console.log(details.status);
  };

  return (
    <Avatar.Root onLoadingStatusChange={handleLoadingStatusChange}>
      <Avatar.Fallback>PA</Avatar.Fallback>
      <Avatar.Image src="https://i.pravatar.cc/300" alt="avatar" />
    </Avatar.Root>
  );
};
```

- Added a `Context` component to allow access to the internal machine API. Previously, it was only possible to access the internal API at the root level, which is manageable for small components but could lead to cumbersome composition in larger components. Additionally, this pattern clashed with the `asChild` composition pattern we use.

```tsx
export const Basic = () => (
  <Popover.Root>
    <Popover.Trigger>Open</Popover.Trigger>
    <Popover.Positioner>
      <Popover.Context>
        {(api) => (
          <Popover.Content>
            <Popover.Title onClick={() => api.close()}>Title</Popover.Title>
            <Popover.Description>Description</Popover.Description>
          </Popover.Content>
        )}
      </Popover.Context>
    </Popover.Positioner>
  </Popover.Root>
);
```

### Removed

- BREAKING: Removed the option to access the internal API from various Root components. Use the new `Context` component instead. This change will help in streamlining the `asChild` composition pattern.

**Comment by gurkerl83** [comment]
Hi, maybe a few other things could be considered doing

Remove the package `@ark-ui/anatomy` and soly rely on @zag/anatomy, however the extend and rename mechanism has to be managed differently.

Also streamline the export and re-export strategy. Here tests have to be conducted, esp. in the sense of barrel files, normally bundlers have a hard time optimizing things.

here different exports are used

- export * as Collapsible from './collapsible' within other parts are collected and exported
- then in the following these parts are exported again

```
import type { OpenChangeDetails as CollapsibleOpenChangeDetails } from '@zag-js/collapsible'
import { CollapsibleContent, type CollapsibleContentProps } from './collapsible-content'
import { useCollapsibleContext, type CollapsibleContext } from './collapsible-context'
import { CollapsibleRoot, type CollapsibleRootProps } from './collapsible-root'
import { CollapsibleTrigger, type CollapsibleTriggerProps } from './collapsible-trigger'

export * as Collapsible from './collapsible'

export { CollapsibleContent, CollapsibleRoot, CollapsibleTrigger, useCollapsibleContext }

export type {
  CollapsibleContentProps,
  CollapsibleContext,
  CollapsibleOpenChangeDetails,
  CollapsibleRootProps,
  CollapsibleTriggerProps,
}
```

Thx!

**Comment by cschroeter** [comment]
@gurkerl83

Thanks for taking the time to comment on this. I have some follow-up questions for you.

1.) I struggle to see the benefits of removing `@ark-ui/anatomy` in favor of exclusively using `@zag/anatomy`. In my view, anatomy is a cross-cutting concern for all components across all frameworks. The only improvement I foresee is exporting the anatomy for the individual components directly within the components themselves.

2.) I'm not convinced that we should export the files twice. The reason for the dual exports is to allow for both types of imports, e.g., `AccordionItem` and `Accordion.Item`. I would prefer to exclusively use the `Accordion.Item` variant across all components, but this might be a bit too opinionated.

We tested this setup with modern bundlers like Vite, Next, and ESBuild and did not spot any issues. What would your proposal be?

**Comment by gurkerl83** [comment]
> @cschroeter 
> 2.) I'm not convinced that we should export the files twice. The reason for the dual exports is to allow for both types of imports, e.g., `AccordionItem` and `Accordion.Item`. I would prefer to exclusively use the `Accordion.Item` variant across all components, but this might be a bit too opinionated.

In the current source components are exported twice.

E.g. in the accordion components index file by

```
export {
  AccordionItem,
  AccordionItemContent,
  AccordionItemIndicator,
  AccordionItemTrigger,
  AccordionRoot,
  useAccordionContext,
  useAccordionItemContext,
}
```

and another time by 

```
export * as Accordion from './accordion' 
```

Thx!

**Comment by cschroeter** [comment]
@gurkerl83 

Got it. This is actually desired. Other headless component libraries do the same (https://github.com/radix-ui/primitives/blob/main/packages/react/accordion/src/index.ts)

If you have a clear reason or benefit why we should not do this. Please let me know :)

**Comment by gurkerl83** [comment]
@cschroeter The radix-ui exports are different, here not export * from Accordion is involved.

In the ark accordion package the result doing this

```
export * as Accordion from './accordion' 
```

is which is fine.

```
Accordion.Item
```

Also the index where `export * as Accordion from './accordion'` is specified AccordionItem and all the other components are re-exported which seems unnecessary when your goal is to access the item component like this `Accordion.Item`

My suggestion would be to change the index file of accordion to something like

here you can work types from accordion, this type export is probably required because e.g. park is using them, but most important the component of a specific component segment is not exported only its type.

```
// Namespace export for grouped components from 'accordion.ts'
export * as Accordion from './accordion';

// Direct exports for unique hooks
export { useAccordionContext } from './accordion-context';
export { useAccordionItemContext } from './accordion-item-context';

// Direct re-export of types for external use
export type { AccordionContext } from './accordion-context';
export type { AccordionItemContext } from './accordion-item-context';
export type { AccordionItemProps } from './accordion-item';
export type { AccordionItemContentProps } from './accordion-item-content';
export type { AccordionItemIndicatorProps } from './accordion-item-indicator';
export type { AccordionItemTriggerProps } from './accordion-item-trigger';
export type { AccordionRootProps } from './accordion-root';

// Importing and then re-exporting types from '@zag-js/accordion' with aliases
import type {
  FocusChangeDetails as AccordionFocusChangeDetails,
  ValueChangeDetails as AccordionValueChangeDetails,
} from '@zag-js/accordion';

export type {
  AccordionFocusChangeDetails,
  AccordionValueChangeDetails,
};
```

Thx!

**Comment by gurkerl83** [comment]
About my first proposal to remove @ark-ui/anatomy... It is not immediately clear and I have not spent enough time why the extend and rename mechanism for some atoms is the e.g. https://github.com/chakra-ui/ark/blob/main/packages/anatomy/src/carousel/index.ts but not e.g. in here https://github.com/chakra-ui/ark/blob/main/packages/anatomy/src/accordion/index.ts For most of them there is just none of those actions. It would be better to put these modifications at the relevant places where these things are immediately used. However I understand the idea of having them in a central place, but maybe these can be moved to zag if it is not a pure component specific feature.

Thx!

**Comment by gurkerl83** [comment]
There is another thingy to consider especially with all the change how context / api is exposed now. 

E.g.

```
export const Basic = () => (
  <Popover.Root>
    <Popover.Trigger>Open</Popover.Trigger>
    <Popover.Positioner>
      <Popover.Context> // 1.) needs to be a client component from here downwards 
        {(api) => (
          <Popover.Content>
            <Popover.Title onClick={() => api.close()}>Title</Popover.Title>
            <Popover.Description>Description</Popover.Description>
          </Popover.Content>
        )}
      </Popover.Context>
    </Popover.Positioner>
  </Popover.Root>
);
```

About comment 1) in code
Using context which is in the latest React versions (canary) not available on server anymore using api / context subcomponents needs to be especially marked as client components.

One thing however what I do not clearly understand, this can be a server component, at least when using pure ark component. It seems when not using render prop based api declaration everything can be in fact a server component (please do not confuse this with SSR).

The following worked out of the box with not 'use client' declaration at all.

```
<Accordion.Root defaultValue={['React']}>
   <>
      {['React', 'Solid', 'Vue'].map((item, id) => (
         <Accordion.Item key={id} value={item}>
         <Accordion.ItemTrigger>
            What is {item}?
            <Accordion.ItemIndicator>
               <ChevronDownIcon />
            </Accordion.ItemIndicator>
         </Accordion.ItemTrigger>
         <Accordion.ItemContent>
            {item} is a JavaScript library for building user interfaces.
              </Accordion.ItemContent>
          </Accordion.Item>
         ))}
   </>
</Accordion.Root>
```

Can we make sure the change towards context exposure does access create / access context by accident when we dont want the api, so we keep e.g. the last example as a full server component.

Thx!

**Comment by cschroeter** [comment]
@gurkerl83 

> There is another thingy to consider especially with all the change how context / api is exposed now.

This is not new behavior. In `v2`, most of the components exposed the underlying machine API in their root components. For example:

```tsx
<Popover.Root>
  {(api) => ()
</Popover.Root>
```

We are trying to address this duplication in version 3 with the help of context components.

```tsx
<Popover.Root>
  <Popover.Context>
    {(api) => ()
  </Popover.Context>
</Popover.Root>
```

This has multiple benefits, especially when looking at the asChild composition.

> Can we make sure the change towards context exposure does access create / access context by accident when we do not require the api, so we keep e.g. the last example as a full server component.

Accessing the internal machine is, in most cases, optional and is just a way for developers to tinker with the internals if needed.

Please be aware that Ark UI, and many other headless component libraries, require access to the DOM, the `window` object, and more. Therefore, it is entirely reasonable to view them as client components. In my opinion, React, and especially Next.js, have done a poor job of educating developers on what `use client` actually means and that it is not inherently bad.

**Comment by gurkerl83** [comment]
Access to DOM etc I sure understand that client is necessary, and I agree the `use client` declaration gets misused quite often these days even in popular ui-libraries. I observed this a few times and for the most part these people decide for a quick fix, this declaration should be used as a last resort controllable by the consumer, maybe in a wrapper to change the behavior but not in a ui-library itself. 

Maybe I`m wrong but the other thing about DOM access is not necessary at the outer shells of a composed component but rather on the leaves but this may depend. Previously my concern was only related to the context API which was removed from react canary a few month ago. 

Allowing the consumer however to decide when and on which level of a component to consume the context is a better idea not requiring doing everything on the root of a component.

Thx!
